File: CS/WindowsFormsApplication1/DragDropResourcesHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraScheduler;
using System.Collections;
using DevExpress.XtraScheduler.Drawing;
using System.Windows.Forms;
using System.Drawing;
using DevExpress.Services;

namespace WindowsFormsApplication1 {
    public class DragDropResourcesHelper {
        private SchedulerControl CurrentScheduler;
        private string CustomFieldName;

        public DragDropResourcesHelper(SchedulerControl scheduler, string fieldName) {
            CurrentScheduler = scheduler;
            CustomFieldName = fieldName;

            CurrentScheduler.Storage.ResourceCollectionLoaded += new EventHandler(Storage_ResourceCollectionLoaded);
            CurrentScheduler.DragOver += new System.Windows.Forms.DragEventHandler(CurrentScheduler_DragOver);
            CurrentScheduler.DragDrop += new System.Windows.Forms.DragEventHandler(CurrentScheduler_DragDrop);

            IMouseHandlerService oldMouseHandler = (IMouseHandlerService)CurrentScheduler.GetService(typeof(IMouseHandlerService));
            if(oldMouseHandler != null) {
                DragrDropMouseHandlerService newMouseHandler = new DragrDropMouseHandlerService(CurrentScheduler, oldMouseHandler);
                CurrentScheduler.RemoveService(typeof(IMouseHandlerService));
                CurrentScheduler.AddService(typeof(IMouseHandlerService), newMouseHandler);
            }
        }

        void CurrentScheduler_DragDrop(object sender, System.Windows.Forms.DragEventArgs e) {
            if(e.Data.GetDataPresent(typeof(DevExpress.XtraScheduler.Resource))) {

                SchedulerHitInfo dropHitInfo = CurrentScheduler.ActiveView.ViewInfo.CalcHitInfo(CurrentScheduler.PointToClient(new Point(e.X, e.Y)), false);
                if(dropHitInfo.HitTest == SchedulerHitTest.ResourceHeader) {
                    Resource sourceResource = e.Data.GetData(typeof(DevExpress.XtraScheduler.Resource)) as DevExpress.XtraScheduler.Resource;
                    Resource targetResource = dropHitInfo.ViewInfo.Resource;
                    if(sourceResource != targetResource) {
                        object sourceResourceSortOrder = sourceResource.CustomFields[CustomFieldName];
                        sourceResource.CustomFields[CustomFieldName] = targetResource.CustomFields[CustomFieldName];
                        targetResource.CustomFields[CustomFieldName] = sourceResourceSortOrder;
                        ApplySorting();
                    }
                }
            }
        }

        void CurrentScheduler_DragOver(object sender, System.Windows.Forms.DragEventArgs e) {
            if(e.Data.GetDataPresent(typeof(DevExpress.XtraScheduler.Resource))) {
                e.Effect = DragDropEffects.Move;
            }
        }

        void ApplySorting() {
            CurrentScheduler.Storage.Resources.Items.Sort(new ResourceCustomFieldComparer("SortOrder"));
            CurrentScheduler.ActiveView.LayoutChanged();
        }


        void Storage_ResourceCollectionLoaded(object sender, EventArgs e) {
            ApplySorting();
        }
    }

    // MOUSE HANDLER SERVICE
    public class DragrDropMouseHandlerService : MouseHandlerServiceWrapper {
        IServiceProvider provider;
        SchedulerHitInfo downHitInfo;
        Point downHitPoint;


        public DragrDropMouseHandlerService(IServiceProvider provider, IMouseHandlerService service)
            : base(service) {
            this.provider = provider;
        }

        public override void OnMouseDown(MouseEventArgs e) {
            downHitInfo = null;
            downHitPoint = Point.Empty;
            SchedulerHitInfo hitInfo = (provider as SchedulerControl).ActiveView.ViewInfo.CalcHitInfo(e.Location, false);
            if(hitInfo.HitTest == SchedulerHitTest.ResourceHeader) {
                downHitInfo = hitInfo;
                downHitPoint = e.Location;
            }
            else 
                base.OnMouseDown(e);
        }

        public override void OnMouseMove(MouseEventArgs e) {
            if(e.Button == MouseButtons.Left && downHitInfo != null) {
                Size dragSize = SystemInformation.DragSize;
                Rectangle dragRect = new Rectangle(new Point(downHitPoint.X - dragSize.Width / 2, downHitPoint.Y - dragSize.Height / 2), dragSize);

                if(!dragRect.Contains(new Point(e.X, e.Y))) {
                    (provider as SchedulerControl).DoDragDrop(downHitInfo.ViewInfo.Resource, DragDropEffects.All);
                    downHitInfo = null;
                    downHitPoint = Point.Empty;
                }
            }
            else
                base.OnMouseMove(e);            
        }
    }



    // RESOURECE COMPARER
    public abstract class ResourceBaseComparer : IComparer<Resource>, IComparer {
        #region IComparer Members
        int IComparer.Compare(object x, object y) {
            return CompareCore(x, y);
        }
        public int Compare(Resource x, Resource y) {
            return CompareCore(x, y);
        }
        #endregion

        protected virtual int CompareCore(object x, object y) {
            Resource xRes = (Resource)x;
            Resource yRes = (Resource)y;

            if(xRes == null || yRes == null)
                return 0;
            if(Object.Equals(xRes.Id, ResourceEmpty.Id) || Object.Equals(yRes.Id, ResourceEmpty.Id))
                return 0;

            return CompareResources(xRes, yRes);
        }

        protected abstract int CompareResources(Resource xRes, Resource yRes);
    }

    public class ResourceCustomFieldComparer : ResourceBaseComparer {
        string customFieldValue = "";
        public ResourceCustomFieldComparer(string parCustomField) {
            customFieldValue = parCustomField;
        }

        protected override int CompareResources(Resource xRes, Resource yRes) {
            return Convert.ToInt32(xRes.CustomFields[customFieldValue]).CompareTo(Convert.ToInt32(yRes.CustomFields[customFieldValue]));
        }
    }
}

File: VB/WindowsFormsApplication1/DragDropResourcesHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraScheduler
Imports System.Collections
Imports DevExpress.XtraScheduler.Drawing
Imports System.Windows.Forms
Imports System.Drawing
Imports DevExpress.Services

Namespace WindowsFormsApplication1
	Public Class DragDropResourcesHelper
		Private CurrentScheduler As SchedulerControl
		Private CustomFieldName As String

		Public Sub New(ByVal scheduler As SchedulerControl, ByVal fieldName As String)
			CurrentScheduler = scheduler
			CustomFieldName = fieldName

			AddHandler CurrentScheduler.Storage.ResourceCollectionLoaded, AddressOf Storage_ResourceCollectionLoaded
			AddHandler CurrentScheduler.DragOver, AddressOf CurrentScheduler_DragOver
			AddHandler CurrentScheduler.DragDrop, AddressOf CurrentScheduler_DragDrop

			Dim oldMouseHandler As IMouseHandlerService = CType(CurrentScheduler.GetService(GetType(IMouseHandlerService)), IMouseHandlerService)
			If oldMouseHandler IsNot Nothing Then
				Dim newMouseHandler As New DragrDropMouseHandlerService(CurrentScheduler, oldMouseHandler)
				CurrentScheduler.RemoveService(GetType(IMouseHandlerService))
				CurrentScheduler.AddService(GetType(IMouseHandlerService), newMouseHandler)
			End If
		End Sub

		Private Sub CurrentScheduler_DragDrop(ByVal sender As Object, ByVal e As System.Windows.Forms.DragEventArgs)
			If e.Data.GetDataPresent(GetType(DevExpress.XtraScheduler.Resource)) Then

				Dim dropHitInfo As SchedulerHitInfo = CurrentScheduler.ActiveView.ViewInfo.CalcHitInfo(CurrentScheduler.PointToClient(New Point(e.X, e.Y)), False)
				If dropHitInfo.HitTest = SchedulerHitTest.ResourceHeader Then
					Dim sourceResource As Resource = TryCast(e.Data.GetData(GetType(DevExpress.XtraScheduler.Resource)), DevExpress.XtraScheduler.Resource)
					Dim targetResource As Resource = dropHitInfo.ViewInfo.Resource
					If sourceResource IsNot targetResource Then
						Dim sourceResourceSortOrder As Object = sourceResource.CustomFields(CustomFieldName)
						sourceResource.CustomFields(CustomFieldName) = targetResource.CustomFields(CustomFieldName)
						targetResource.CustomFields(CustomFieldName) = sourceResourceSortOrder
						ApplySorting()
					End If
				End If
			End If
		End Sub

		Private Sub CurrentScheduler_DragOver(ByVal sender As Object, ByVal e As System.Windows.Forms.DragEventArgs)
			If e.Data.GetDataPresent(GetType(DevExpress.XtraScheduler.Resource)) Then
                e.Effect = DragDropEffects.Move
			End If
		End Sub

		Private Sub ApplySorting()
			CurrentScheduler.Storage.Resources.Items.Sort(New ResourceCustomFieldComparer("SortOrder"))
			CurrentScheduler.ActiveView.LayoutChanged()
		End Sub


		Private Sub Storage_ResourceCollectionLoaded(ByVal sender As Object, ByVal e As EventArgs)
			ApplySorting()
		End Sub
	End Class

	' MOUSE HANDLER SERVICE
	Public Class DragrDropMouseHandlerService
		Inherits MouseHandlerServiceWrapper
		Private provider As IServiceProvider
		Private downHitInfo As SchedulerHitInfo
		Private downHitPoint As Point


		Public Sub New(ByVal provider As IServiceProvider, ByVal service As IMouseHandlerService)
			MyBase.New(service)
			Me.provider = provider
		End Sub

		Public Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			downHitInfo = Nothing
			downHitPoint = Point.Empty
			Dim hitInfo As SchedulerHitInfo = (TryCast(provider, SchedulerControl)).ActiveView.ViewInfo.CalcHitInfo(e.Location, False)
			If hitInfo.HitTest = SchedulerHitTest.ResourceHeader Then
				downHitInfo = hitInfo
				downHitPoint = e.Location
			Else
				MyBase.OnMouseDown(e)
			End If
		End Sub

		Public Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			If e.Button = MouseButtons.Left AndAlso downHitInfo IsNot Nothing Then
				Dim dragSize As Size = SystemInformation.DragSize
				Dim dragRect As New Rectangle(New Point(downHitPoint.X - dragSize.Width / 2, downHitPoint.Y - dragSize.Height / 2), dragSize)

				If (Not dragRect.Contains(New Point(e.X, e.Y))) Then
					TryCast(provider, SchedulerControl).DoDragDrop(downHitInfo.ViewInfo.Resource, DragDropEffects.All)
					downHitInfo = Nothing
					downHitPoint = Point.Empty
				End If
			Else
				MyBase.OnMouseMove(e)
			End If
		End Sub
	End Class



	' RESOURECE COMPARER
	Public MustInherit Class ResourceBaseComparer
		Implements IComparer(Of Resource), IComparer
		#Region "IComparer Members"
		Private Function IComparer_Compare(ByVal x As Object, ByVal y As Object) As Integer Implements IComparer.Compare
			Return CompareCore(x, y)
		End Function
		Public Function Compare(ByVal x As Resource, ByVal y As Resource) As Integer Implements IComparer(Of Resource).Compare
			Return CompareCore(x, y)
		End Function
		#End Region

		Protected Overridable Function CompareCore(ByVal x As Object, ByVal y As Object) As Integer
			Dim xRes As Resource = CType(x, Resource)
			Dim yRes As Resource = CType(y, Resource)

			If xRes Is Nothing OrElse yRes Is Nothing Then
				Return 0
			End If
			If Resource.Empty.Equals(xRes) OrElse Resource.Empty.Equals(yRes) Then
				Return 0
			End If

			Return CompareResources(xRes, yRes)
		End Function

		Protected MustOverride Function CompareResources(ByVal xRes As Resource, ByVal yRes As Resource) As Integer
	End Class

	Public Class ResourceCustomFieldComparer
		Inherits ResourceBaseComparer
		Private customFieldValue As String = ""
		Public Sub New(ByVal parCustomField As String)
			customFieldValue = parCustomField
		End Sub

		Protected Overrides Function CompareResources(ByVal xRes As Resource, ByVal yRes As Resource) As Integer
            Return Convert.ToInt32(xRes.CustomFields(customFieldValue)).CompareTo(Convert.ToInt32(yRes.CustomFields(customFieldValue)))
		End Function
	End Class
End Namespace
